### DISCUSSION TITLE
Drop Python 3.7 support

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#2418] [General]
Similar to https://github.com/py-pdf/pypdf/discussions/2005, I want to have this topic open to see the pros and cons of dropping Python 3.7 support.

This discussion will remain open until the reasons to drop 3.7 become stronger than the reasons to keep the support.

If you have an opinion on the topic, you can vote on one of the two comments of this post: drop support / keep support.

**Comment by MartinThoma** [comment]
Drop Python 3.7 support reasons:

* [Python 3.7 reached it's end-of-live](https://peps.python.org/pep-0537/) in June 2023
* Important packages dropped 3.7 support:
  * [pre-commit==3.0.0](https://github.com/pre-commit/pre-commit/blob/main/CHANGELOG.md) dropped Python<3.8 support in January 2023
* Comparable packages that dropped 3.7 support:
  * [`pymupdf==1.23.0`](https://pymupdf.readthedocs.io/en/latest/changes.html) dropped Python<3.8 support in August 2023
  * [`pikepdf==7.0.0`](https://pikepdf.readthedocs.io/en/latest/releasenotes/version7.html#v7-0-0) dropped Python<3.8 in February 2023
* OS default versions:
   * [Ubuntu 20.04](https://askubuntu.com/a/1232820) has 3.8.2. That is the [oldes Ubuntu version](https://en.wikipedia.org/wiki/Ubuntu_version_history) still stupported
* [Killer features of newer Python versions](https://python.plainenglish.io/killer-features-by-python-version-c84ca12dba8):
  * [PEP 589](https://peps.python.org/pep-0589/): TypedDict - I'm unsure if we want to use that, though. In my experience they are too rigid

**Reply by stefan6419846** [reply]
I am voting for *pypdf* syncing with the regular Python EOL timelines as far as possible. This avoids dealing with possible dependency conflicts and ideally allows us to use the same package versions for most of the requirements without having to look at the Python version.

**Reply by pubpub-zz** [reply]
some linux LTS distributions still active  are using old python versions (eg Debian 9 "Buster" is in LTS up to June and references Python 3.7)
so for dropping 3.7 I would wait summer time.
For me, as long as the effort to is limited, we should try to keep support

**Comment by MartinThoma** [comment]
Keep Python 3.7 support

* [pypdf downloads](https://pypistats.org/packages/pypdf):
  * **3.19% use Python 3.7**
  * 8.26% have no version specified
  * 0.4% use versions that we don't support anymore
* OS default Python versions
   * [Debian 10](https://linuxize.com/post/how-to-install-python-3-9-on-debian-10/) has Python 3.7. That is the [oldest Debian version still supported](https://en.wikipedia.org/wiki/Debian_version_history)  until **June 2024**
* Related packages
  * [fpdf2==2.7.x](https://pypi.org/project/fpdf2/) still supports Python 3.7

**Comment by pubpub-zz** [comment]
I add another opinion : Don't mind 😜

**Comment by stefan6419846** [comment]
Python 3.7 reached its end of life over a year ago, Debian 10 just reached its EOL as well.

The largest benefit of dropping Python 3.7 would be to use more recent package versions in CI more easily as discussed in #2455 and #2574.

**Comment by shartzog** [comment]
I am the walrus (:=)...
That was the biggest deal for 3.8 imo... ;)

**Comment by shartzog** [comment]
The assignment operator (:=) opens up lots of performance improvements for comprehensions, lambdas and `any()` conditionals.
